spec/mongo/transactions_examples_spec.rb

```diff
@@ -0,0 +1,236 @@
+require 'spec_helper'
+
+describe 'transactions examples in Ruby', if: test_transactions? do
+
+  let(:hr) do
+    authorized_client.use(:hr).database
+  end
+
+  let(:reporting) do
+    authorized_client.use(:reporting).database
+  end
+
+  before(:each) do
+    hr[:employees].insert_one(employee: 3, status: 'Active')
+    reporting[:events].insert_one(employee: 3, status: { new: 'Active', old: nil})
+  end
+
+  after(:each) do
+    hr.drop
+    reporting.drop
+  end
+
+  context 'individual examples' do
+
+    let(:session) do
+      authorized_client.start_session
+    end
+
+    # Start Transactions Intro Example 1
+
+    def update_employee_info(session)
+      employees_coll = session.client.use(:hr)[:employees]
+      events_coll = session.client.use(:reporting)[:events]
+
+      session.start_transaction(read_concern: { level: :snapshot },
+                                write_concern: { w: :majority })
+      employees_coll.update_one({ employee: 3 }, { '$set' => { status: 'Inactive'} },
+                                session: session)
+      events_coll.insert_one({ employee: 3, status: { new: 'Inactive', old: 'Active' } },
+                             session: session)
+
+      loop do
+        begin
+          session.commit_transaction
+          puts 'Transaction committed.'
+          break
+        rescue Mongo::Error::OperationFailure, Mongo::Error::SocketError,
+               Mongo::Error::NoServerAvailable => e
+
+          if e.label?(Mongo::Session::UNKNOWN_TRANSACTION_COMMIT_LABEL)
+            puts "#{Mongo::Session::UNKNOWN_TRANSACTION_COMMIT_LABEL}, retrying commit operation..."
+            next
+          else
+            puts 'Error during commit ...'
+            raise
+          end
+        end
+      end
+    end
+
+    # End Transactions Intro Example  1
+
+    context 'Transactions Intro Example 1' do
+
+      let(:run_transaction) do
+        update_employee_info(session)
+      end
+
+      it 'makes the changes to the database' do
+        run_transaction
+        employee = hr[:employees].find({ employee: 3 }, limit: 1).first
+        expect(employee).to_not be_nil
+        expect(employee['status']).to eq('Inactive')
+      end
+    end
+
+    context 'Transactions Retry Example 1' do
+
+      # Start Transactions Retry Example 1
+
+      def run_transaction_with_retry(session)
+        loop do
+          begin
+            yield session # performs transaction
+            break
+          rescue Mongo::Error::OperationFailure, Mongo::Error::SocketError,
+                 Mongo::Error::NoServerAvailable => e
```

Comment:
Can `Mongo::Error` be rescued here instead?

Reply:
No; not all error types have labels

Reply:
How does the user know which exceptions have labels and which don't? How was this set of exceptions determined?